File: LinkedList/143. Reorder List/Solution.java
// Definition for singly-linked list.
public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public void reorderList(ListNode head) {
        // N : no of nodes
        // Time Complexity: O(N)
        // Space Complexity: O(1)
        if (head.next == null) return;


        // Finding the left and right portion of the list using slow and fast approach
        ListNode dummy = new ListNode(0, head);
        ListNode left = head;
        ListNode right = dummy;
        ListNode curr = head;
        while (curr != null && curr.next != null) {
            curr = curr.next.next;
            right = right.next;
        }
        curr = right;
        right = right.next;
        curr.next = null;

        // Reverse the right list
        dummy = new ListNode(0, new ListNode(0, right));
        ListNode node1 = dummy;
        ListNode node2 = dummy.next;
        curr = dummy.next.next;
        while (curr != null) {
            node2.next = node1;

            node1 = node2;
            node2 = curr;
            curr = curr.next;
        }
        node2.next = node1;
        right.next = null;
        right = node2;

        // Reorder the head list
        head = curr = left;
        left = left.next;
        while (left != null) {
            curr.next = right;
            right = right.next;
            curr.next.next = left;
            left = left.next;
            curr = curr.next.next;
        }
        curr.next = right;
    }
}

File: LinkedList/143. Reorder List/Solution.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # N : no of nodes
        # Time Complexity: O(N)
        # Space Complexity: O(1)
        if not head.next:
            return
    
        # Finding the left and right portion of the list using slow and fast approach
        dummy = ListNode(0, head)
        left = fast = head
        right = dummy

        while fast and fast.next:
            fast = fast.next.next
            right = right.next

        temp = right
        right = right.next
        temp.next = None

        # Reverse the right list
        dummy = ListNode(0, ListNode(0, right))
        node1, node2, node3 = right, dummy.next, dummy
        while node1:
            node2.next = node3

            node3 = node2
            node2 = node1
            node1 = node1.next
        
        node2.next = node3
        right.next = None
        right = node2

        # Reorder the head list
        head = curr = left
        left = left.next
        while left:
            curr.next = right
            right = right.next
            curr = curr.next
            curr.next = left
            left = left.next
            curr = curr.next
        curr.next = right